components/brave_rewards/core/contribution/contribution.cc

```diff
@@ -288,19 +305,56 @@ void Contribution::OnNextMonthlyContributionTimeRead(
   BLOG(1, "Monthly contribution timer set for " << delay);
 }
 
-void Contribution::SetMonthlyContribution(
-    const std::string& publisher_id,
-    double amount,
-    base::OnceCallback<void(bool)> callback) {
-  ledger_->database()->SetMonthlyContribution(
-      publisher_id, amount,
-      base::BindOnce(&Contribution::OnMonthlyContributionSet,
-                     base::Unretained(this), std::move(callback)));
+void Contribution::SendContribution(const std::string& publisher_id,
+                                    double amount,
+                                    bool set_monthly,
+                                    base::OnceCallback<void(bool)> callback) {
+  ContributionRequest request(publisher_id, amount, set_monthly,
+                              std::move(callback));
+
+  tip_->Process(publisher_id, amount,
+                base::BindOnce(&Contribution::OnContributionRequestQueued,
+                               base::Unretained(this), std::move(request)));
+}
+
+void Contribution::OnContributionRequestQueued(
+    ContributionRequest request,
+    absl::optional<std::string> queue_id) {
+  if (!queue_id) {
+    std::move(request.callback).Run(false);
+    return;
+  }
+
+  DCHECK(!queue_id->empty());
+  requests_.emplace(std::pair(std::move(*queue_id), std::move(request)));
+}
+
+void Contribution::OnContributionRequestCompleted(const std::string& queue_id,
+                                                  bool success) {
+  auto request_iter = requests_.find(queue_id);
+  if (request_iter == requests_.end()) {
+    return;
+  }
+
+  ContributionRequest request(std::move(request_iter->second));
+  requests_.erase(request_iter);
+
+  // If the contribution was successful and the user has requested that this be
+  // a recurring contribution, record the monthly contribution in the database.
+  // Optimistically assume that a write failure will not occur. The callback
+  // should receive the result of the contribution, regardless of whether this
+  // write succeeds or fails.
+  if (success && request.set_monthly) {
+    ledger_->database()->SetMonthlyContribution(
```

Comment:
Shouldn't we consider this write as part of the flow, that is, bind `request.callback` to a parameter of `Contribution::OnMonthlyContributionSet()` and call it with `success` when it finished?

Reply:
Consider the case where the tip succeeds and the write fails:

* The tip has gone through.
* The user wanted a monthly contribution set up for later, but that didn't happen.
* The user sees a message stating something like "your contribution failed, please try again".

I think we want to avoid that situation.

We could send extra information back to the client. Instead of a `bool`, we could have an enum representing:

* tip failed
* tip sent, monthly set failed
* every succeeded

This pushes more responsibility to the client for handling the different situations appropriately. Since a DB write failure would (a) be quite uncommon, and (b) would likely not be "fixable" by the user, it's arguably not worth pushing the added complexity onto the API user or front-end. We can just log the error (I think it should already be logged by the database code?).

Reply:
Hmm... AFAIK, when setting up a monthly contribution, we don't make the first tip right away.